stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -1507,6 +1511,69 @@ func _TFCOpSetAttrStringArray(_ op: CTFEOp,
   }
 }
 
+/// A TensorFlow device kind.
+public enum DeviceKind : UInt {
+  /// Default device.
+  case `default` = 0
+  /// Central processing units.
+  case cpu = 1
+  /// Graphics processing units.
+  case gpu = 2
+  // TODO: TPU?
+}
+
+internal extension _ExecutionContext {
+  var currentDeviceKind: DeviceKind {
+    return sync {
+      deviceScopes.last ?? .default
+    }
+  }
+
+  @usableFromInline
+  func pushDeviceScope(_ kind: DeviceKind) {
+    sync {
+      deviceScopes.append(kind)
+    }
+  }
+
+  @usableFromInline
+  func popDeviceScope() {
+    sync {
+      internalConsistencyCheck(deviceScopes.popLast() != nil)
+    }
+  }
+}
+
+/// Execute a closure, making TensorFlow operations run on a specific kind of
+/// device unless specified otherwise.
+///
+/// - Parameters:
+///   - kind: A kind of device to run TensorFlow operations on.
+///   - body: A closure whose TensorFlow operations are to be executed on the
+///   specified kind of device.
+@inlinable
+public func withDevice<R>(_ kind: DeviceKind, perform body: () throws -> R) rethrows -> R {
+  _ExecutionContext.global.pushDeviceScope(kind)
+  print("push: ", kind)
```

Comment:
The runtime shouldn't print random debugging info to stdout. This should be switched to using `debugLog(_:)` with a descriptive message, e.g. `debugLog("Pushing device \(kind) to the device stack")`. Same for other occurrences of `print`.